Add unsigned integer dtypes to PyTorch

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #116810
* #116809
* #116808
* #116807
* #116806
* #116805
* #116804
* #116803
* #116595
* __->__ #116594
* #116693
* #116698

The dtypes are very useless right now (not even fill works), but it makes torch.uint16, uint32 and uint64 available as a dtype.

Towards https://github.com/pytorch/pytorch/issues/58734

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

ezyang committed:
Update on "Add unsigned integer dtypes to PyTorch"

[ghstack-poisoned]

albanD commented on c10/core/ScalarType.h:
```diff
@@ -424,6 +454,10 @@ static inline bool isBitsType(ScalarType t) {
       t == ScalarType::Bits16;
 }
 
+static inline bool isBarebonesUnsignedType(ScalarType t) {
```
Should UInt8 be in here as well?
I guess not given how you use it for the type promotion checks. But this might be a confusing API for our c++ devs?

ezyang replied:
Bare bones here is defined to be "we have only minimal kernel support in traditional C++ eager mode". Since uint8 is grandfathered from Lua days to have lots of kernels, it shouldn't be included.

albanD replied:
What is the plan to document that to make sure user expectations for these are appropriate?

ezyang replied:
The most logical place to document that these are bare bones is in torch.Tensor where we list supported dtypes. Note that the f8 dtypes are not documented right now, so we could also keep it under the radar undocumented until enough stuff is working. I am also half expecting to end up having a select few eager kernels sprout arbitrary unsigned support, especially gather-like operations I think is what people are most likely to want.

albanD replied:
One could argue that fp8 dtypes are not very well documented and shouldn't be the example here. Also there is a strong alignment there that we will not implement any arithmetic op for these dtypes (no scale means no compute with the value).
These dtypes could contain the full thing so we most likely want to have some details about that?

ezyang replied:
I'll post a proposed doc PR and we can discuss it there

ezyang replied:
https://github.com/pytorch/pytorch/pull/116804

albanD commented on c10/core/ScalarType.h:
```diff
@@ -592,6 +626,12 @@ static inline ScalarType promoteTypes(ScalarType a, ScalarType b) {
   }
 
+  if (isBarebonesUnsignedType(a) && a == b) {
+    return a;
```
Note that the current type promotion is such that if a == b, return a is fine. I expect this to remain true so we might want to refactor that check at the top for minimal speed bump (since it is the common case) and simplifying some of these conditions.

ezyang replied:
Fixed post Driss rebase.

albanD reviewed (approved):
nit on the c++ API naming but sounds good.
it might be good to collect a full list on the issue of the todos for these dtype (python binding, op support, testing support, serialization, etc) so that we can ensure we have a clear "current state" for them somewhere.

ezyang committed:
Update

[ghstack-poisoned]

vadimkantorov commented:
a new era for pytorch in 2024! :)

ezyang committed:
Update

[ghstack-poisoned]

ezyang committed:
Update on "Add unsigned integer dtypes to PyTorch"


The dtypes are very useless right now (not even fill works), but it makes torch.uint16, uint32 and uint64 available as a dtype.

Towards https://github.com/pytorch/pytorch/issues/58734

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]

ezyang committed:
Update on "Add unsigned integer dtypes to PyTorch"


The dtypes are very useless right now (not even fill works), but it makes torch.uint16, uint32 and uint64 available as a dtype.

Towards https://github.com/pytorch/pytorch/issues/58734

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]

ezyang committed:
Update on "Add unsigned integer dtypes to PyTorch"


The dtypes are very useless right now (not even fill works), but it makes torch.uint16, uint32 and uint64 available as a dtype.

Towards https://github.com/pytorch/pytorch/issues/58734

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]

ezyang commented:
@pytorchbot merge

vadimkantorov commented:
Will reinterpret int16_tensor.view(torch.uint16) and reverse be supported out of the box?

Another useful thing to support for unsigned tensors are more bit ops: https://github.com/pytorch/pytorch/issues/105465 as more bit ops are clearly defined for unsigned types and do not need any special behavior for negative operands

ezyang commented:
Yeah, I didn't test it but the mechanism is dtype agnostic.

We probably will end up negotiating some sort of basic set of ops which we will provide in eager mode, I'm thinking of letting the current binary size changes settle first before moving on.
